### PR TITLE
Intel ADSP: add support for "cold" code and data

### PR DESCRIPTION
**Pull request by lyakh**
On Intel ADSP the firmware is first stored in abundant but slower DRAM before being copied to scarce and fast SRAM. This PR adds support for storing and executing data and code directly in DRAM without copying it to SRAM.

### CODE REVIEW

File: soc/intel/intel_adsp/ace/ace-link.ld [hunk 1]
```diff
@@ -171,6 +171,16 @@ SECTIONS {
    */
   .imrdata : ALIGN(4096) {
     *(.imrdata .imrdata.*)
+  } >imr
+
+  /* Cold code in IMR memory */
+  .cold : ALIGN(4096) {
+    *(.cold .cold.* )
```
Comment (andyross) [inline soc/intel/intel_adsp/ace/ace-link.ld:178]: Seems like this is basically identical to the pre-existing .imrdata and should be unified, no?

**Review by andyross — APPROVED** [review]
Nice.  I'm literally looking at doing something very similar right now for mtk, which has the opposite problem (general .text goes in DRAM, but there needs to be a way to put "hot" stuff into the SRAM sections).

You should probably go into `include/zephyr/linker/section_tags.h` and add a `__cold` attribute to serve as a general API.  Probably unifying that with the pre-existing `__imr` along the way.

One suggestion here is to rename the relevant linker sections to e.g. `.rodata.hot.xxx` and `.text.hot.xxx`.  That way pre-existing linker scripts will continue to work, but the platform has the ability to recognize them and put them somewhere special.  This design is still a little half-baked, to be fair; it works for me locally as a prototype but I don't have a final version.

**Comment by lyakh** [comment]
> Nice. I'm literally looking at doing something very similar right now for mtk, which has the opposite problem (general .text goes in DRAM, but there needs to be a way to put "hot" stuff into the SRAM sections).
> 
> You should probably go into `include/zephyr/linker/section_tags.h` and add a `__cold` attribute to serve as a general API. Probably unifying that with the pre-existing `__imr` along the way.
> 
> One suggestion here is to rename the relevant linker sections to e.g. `.rodata.hot.xxx` and `.text.hot.xxx`. That way pre-existing linker scripts will continue to work, but the platform has the ability to recognize them and put them somewhere special. This design is still a little half-baked, to be fair; it works for me locally as a prototype but I don't have a final version.

@andyross I might be wrong, but as I tried, exactly that didn't work for me - using `.text.hot.*` etc. Then it just gets merged into `.text` which is exactly what I have to avoid. Yes, maybe I should define `__cold` next to `__imr` - I was thinking about that, although `__imr` is used for boot code, that does any bootstrapping and copying of "relevant" code to SRAM. As for unifying - yes, I could reuse `__imr` for `__cold` too, but so far we aren't sure if all configurations can handle that, and separating them seems cleaner anyway.

**Comment by andyross** [comment]
Right, it's more a concern that `__imr` is sort of a kludge, added back when I didn't really know what to call it but wanted the ability to put stuff there in the linker.  We should probably consider deprecating and renaming it semantically, e.g. `__boot` /`__early`/`__core` (the last could be shared with `__pinned`, I suspect).  And if it happens to be implemented the same as `__cold` in the platform layer, so much the better.

Basically my concern is that our section_tags.h header is a big mess of special cases and we should start trying to prune it down into simpler hints set on the section names and let the platform layers do the complicated logic.

As as far as naming, it's working for me (barely, in a hacky prototype).  But it's sensitive to ordering in the linker file.  So if you try to do it after your linker.ld includes `common-rom.ld`, then that script will see the `.text.hot.` sections and grab them before you can do anything.

**Review by ceolin — APPROVED** [review]

**Pull request merged**